### PR TITLE
Twitter image falls back to first image in product image gallery

### PR DESCRIPTION
**Pull request by hansjovis**
## Context

* We want the Twitter image to fall back to the first image in the product gallery when no product image is set, to ensure that Twitter shows an image when sharing the webpage.

## Summary

This PR can be summarized in the following changelog entry:

* Fixes a bug where the `twitter:image` meta tag did not fall back to the first product gallery image when no main product image was set.

## Relevant technical choices:

* N/A

## Test instructions

This PR can be tested by following these steps:

### Happy path
* Disable the open graph meta data 
  * (set the 'Add Open Graph meta data' toggle to 'disabled' in _SEO_ > _Social_ > _Facebook_).
* Create a new WooCommerce product.
* Add multiple images to the product image gallery.
* Publish the product.
* Go to the frontend and check the HTML source code.
* The source code should have a `twitter:image` meta tag pointing to the URL of the first image in the product image gallery.

### Should not fall back to product gallery image if main product image is set.
* Go back to the product.
* Add a main product image.
* Go to the frontend and check the HTML source code.
* The source code should have a `twitter:image` meta tag pointing to the URL of the main product image.

### Should not display a twitter image meta tag when open graph is enabled.
* Enable open graph meta data.
* Go to the frontend of the product and check the HTML source code.
* The source code should **not** have a `twitter:image` meta tag.

Fixes https://github.com/Yoast/wpseo-woocommerce/issues/587

### CODE REVIEW

File: classes/woocommerce-twitter.php [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	 * Lets the twitter image fall back to the first image in the product gallery.
+	 *
+	 * @param string                 $twitter_image The current twitter image.
+	 * @param Indexable_Presentation $presentation  The indexable presentation.
+	 *
+	 * @return string The image fallback.
+	 */
+	public function fallback_to_product_gallery_image( $twitter_image, $presentation ) {
+		// We should only fall back to the Twitter image if OpenGraph is disabled and no other Twitter image is set.
+		if ( ! $presentation->context->open_graph_enabled && ! $twitter_image ) {
```
Comment (igorschoester) [inline classes/woocommerce-twitter.php:30]: I would prefer a reverse of this logic so the nesting stays limited.

File: classes/woocommerce-twitter.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+				$product = \wc_get_product( $object->object_id );
+
+				if ( $product ) {
+					$gallery_image_ids      = $product->get_gallery_image_ids();
+					$first_gallery_image_id = \reset( $gallery_image_ids );
+
+					if ( $first_gallery_image_id ) {
+						return YoastSEO()->helpers->twitter->image->get_by_id( $first_gallery_image_id );
+					}
+					return '';
```
Comment (igorschoester) [inline classes/woocommerce-twitter.php:44]: If this is a fallback, why overwrite the image when no image is found?

File: classes/woocommerce-twitter.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	 *
+	 * @return string The image fallback.
+	 */
+	public function fallback_to_product_gallery_image( $twitter_image, $presentation ) {
+		// We should only fall back to the Twitter image if OpenGraph is disabled and no other Twitter image is set.
+		if ( ! $presentation->context->open_graph_enabled && ! $twitter_image ) {
+			$object = $presentation->model;
+
+			// Fall back to the first image in the product gallery.
+			if ( $object->object_type === 'post' && $object->object_sub_type === 'product' ) {
```
Comment (igorschoester) [inline classes/woocommerce-twitter.php:34]: Maybe not nest this either?

**Comment by johannadevos** [comment]
CR: ✅ 
- [x] The comments from the previous CR are addressed.

**Comment by johannadevos** [comment]
Acceptance: ✅

**Pull request merged**